libbeat/publisher/client.go

```diff
@@ -78,32 +79,49 @@ func newClient(pub *PublisherType) *client {
 }
 
 func (c *client) PublishEvent(event common.MapStr, opts ...ClientOption) bool {
-	c.annotateEvent(event)
-
+	publishEvent := c.alterEvent(event)
+	if publishEvent == nil {
+		return false
+	}
 	ctx, client := c.getClient(opts)
 	publishedEvents.Add(1)
-	return client.PublishEvent(ctx, event)
+	return client.PublishEvent(ctx, *publishEvent)
 }
 
 func (c *client) PublishEvents(events []common.MapStr, opts ...ClientOption) bool {
+	publishEvents := []common.MapStr{}
+
 	for _, event := range events {
-		c.annotateEvent(event)
+		publishEvent := c.alterEvent(event)
+		if publishEvent != nil {
+			publishEvents = append(publishEvents, *publishEvent)
+		}
```

Comment:
loop ok. Go idiomatic In-place copy trick (no additional allocations):

    publishEvents := events[:0]
    ...

If publishEvents is empty no need to forward events + trigger Signal within context to indicate success (filtered events should be treated as send successfully).

Doing filtering early plays very well with 'Guaranteed' flag.